src/Common/src/Interop/Windows/Winsock/Interop.WinsockAsync.cs

```diff
@@ -161,14 +165,29 @@ internal struct NetworkEvents
         }
 
         [StructLayout(LayoutKind.Sequential)]
-        internal struct WSAMsg
+        internal unsafe struct WSAMsg
         {
-            internal IntPtr socketAddress;
+            internal byte* socketAddress;
             internal uint addressLength;
-            internal IntPtr buffers;
+            internal WSABuffer* buffers;
             internal uint count;
             internal WSABuffer controlBuffer;
             internal SocketFlags flags;
+            internal int pad;
+
+            // NOTE: Even though Win32 docs say the flags arg is a DWORD, and thus 4 bytes long, it seems to actually write 4 bytes past this.
+            // The "pad" field above deals with this.
```

Comment:
Stepping back:

Yes, this makes me nervous too.

I think it would help a lot if we understood exactly what's going on here.

I wrote some code to test my hypothesis that there's an issue in the WSARecvMsg implementation itself, and couldn't confirm this.  So, most likely this is not the actual problem.

The other possibility is that there's a CLR problem here; I'll write some code to see if I can induce this issue.